### ISSUE TITLE
Why are half my uniforms missing when initializing a compute shader?

### ISSUE THREAD

**Issue by LeoBosse** [moderngl/moderngl#537]
Hello, I've posted this question on stack overflow already ([link](https://stackoverflow.com/questions/74782988/why-are-half-my-uniforms-missing-when-initializing-a-compute-shader-using-python)) but thought I might find more help around here. Don't hesitate to delete if this is not the place.
Here is the question:

Here is the problem: I'm creating a compute shader in python for radiative transfer using modernGL. But upon creation, it does not contain all the uniforms I declare in the shader code. Here is my (simplified) code.
```

# Creation of the context
context = mgl.create_standalone_context(require=440)
# Creation of the compute shader with the source code passed as a string stored in a separate file
compute_shader = context.compute_shader(src_code)

# I set the uniform values stored in a dictionary defined previously. The keys of the dictionary match the uniforms defined in the shader source code.
for key, value in uniforms.items():
    compute_shader.get(key, default=None).write(value)

```
Here is the head of the shader source file 'src_code':
```

#version 440

uniform float instrument_azimut;
uniform float instrument_elevation;
uniform float instrument_altitude;
uniform float instrument_lon;
uniform float instrument_lat;
uniform float instrument_area;
uniform float instrument_fov;
uniform float wavelength;
uniform int   los_nb_points;
uniform float map_delta_az;
uniform bool  is_point_source;
uniform int   atm_nb_altitudes;
uniform int   atm_nb_angles;
uniform bool  use_aerosol;
// Iterable uniforms
uniform float instrument_los[3];
uniform float los_altitudes[100];
uniform float atm_altitudes[121];
uniform float map_azimuts[10];
uniform float map_distances[10];
uniform float los_volumes[100];
uniform float atm_total_abs[121];
uniform float atm_ray_beta[121];
uniform float atm_aer_beta[121];
uniform float atm_angles[100];
uniform float atm_aer_Pfct[121];
uniform float atm_aer_Pfct_DoLP[121];
uniform float los_transmittance[100];


#define X 1
#define Y 1
#define Z 100


// Set up our compute groups
layout(local_size_x = X, local_size_y = Y, local_size_z = Z) in;

// Define constant values
const float EARTH_RADIUS = 6371.; //km
const float PI = 3.14159;
const float DtoR = PI / 180;
const float RtoD = 1. / DtoR;


// Stokes paramter buffers
layout(std430, binding=0) buffer in_0{
    float in_V[];};
layout(std430, binding=1) buffer in_1{
    float in_Vcos[];};
layout(std430, binding=2) buffer in_2{
    float in_Vsin[];};

layout(std430, binding=3) buffer out_3{
    float out_V[];};
layout(std430, binding=4) buffer out_4{
    float out_Vcos[];};
layout(std430, binding=5) buffer out_5{
    float out_Vsin[];};

\\...All useful functions...
```

When I try to list the uniforms contained in the compute shader object, almost half are missing! The following loop
```

for u in compute_shader:
    print(u)

```
outputs:
```

instrument_elevation
instrument_altitude
instrument_area
map_delta_az
is_point_source
atm_nb_altitudes
atm_nb_angles
use_aerosol
los_altitudes
atm_altitudes
map_azimuts
map_distances
los_volumes
atm_total_abs
atm_ray_beta
atm_aer_beta
atm_angles
atm_aer_Pfct
atm_aer_Pfct_DoLP
los_transmittance
```

So when I try to write the value of the uniform "instrument_lon" (float) or "instrument_los" (float array) for example, it doesn't exists in the compute shader object. I'm lost here... I thought about a format issue, so I tried to explicitly set the type of the uniform values in the dictionary. I also tried to declare them in a different order, but it is always the same that are missing. But I can't see any differences with the one initialized correctly...
Do you have any ideas on what I do wrong here?

**Comment by einarf** [comment]
Normal uniforms are stored in a fix sized buffer on the client side and we popualte that buffer when setting the values. When the rendering call is issued this buffer is synced up to gpu behind the scenes. This is a fixed buffer size tha can contain at least 1024 floats/bools/ints (each of them referred to as a "component")

You are clearly exceding this size limit. I think I counted around 1020 components in the uniforms that is actually reported to exist by the compute shader. Uniforms above the limit are usually silently discarded or something might show up in `ctx.error`.

Ref : https://www.khronos.org/opengl/wiki/Uniform_(GLSL)#Implementation_limits

Also some limits are exposed in `ctx.info` (can also be displayed with `python -m moderngl --info`

The correct way to pass such large amounts of data to a compute shader is using SSBOs. These are just buffers containing the data. I would suggest at least making two buffers here containing the arrays with length 100 and 121.

```glsl
// Put these in the first buffer
uniform float los_altitudes[100];
uniform float los_volumes[100];
uniform float atm_angles[100];
uniform float los_transmittance[100];

// Put thise in a second buffer
uniform float atm_altitudes[121];
uniform float atm_total_abs[121];
uniform float atm_ray_beta[121];
uniform float atm_aer_beta[121];
uniform float atm_aer_Pfct[121];
uniform float atm_aer_Pfct_DoLP[121];
```

That will free up 1126 components in the uniform buffer minus 10 for the new ones.

**Comment by LeoBosse** [comment]
Thanks for this detailed answer. I thought I was under the component limit as I found these values in ctx.info:
GL_MAX_COMPUTE_UNIFORM_COMPONENTS = 16384 
GL_MAX_UNIFORM_LOCATIONS = 81920
I also discarded this idea because all the large lists are done correctly, and only the single float and small lists (len=3) are missing. It would mean that I am just over the limit, which seemed unlikely.

I'll try to compact all that in a couple of buffers.

**Comment by einarf** [comment]
Is your `GL_MAX_COMPUTE_UNIFORM_COMPONENTS` really that high? What hardware are you running this on?

**Comment by einarf** [comment]
The second explanation is that the GLSL compiler have determined that the missing uniforms have no effect on the compute shader itself.

**Comment by LeoBosse** [comment]
I have a Dell Inc. Latitude 5420 laptop, with 16GiB of memory, processor 11th Gen Intel® Core™ i7-1165G7 @ 2.80GHz × 8 and a Mesa Intel® Xe Graphics (TGL GT2) graphic card running ubuntu 22.04. Running ```python -m moderngl --info```, I find:
"GL_MAX_COMPUTE_ATOMIC_COUNTERS": 4096,
    "GL_MAX_COMPUTE_ATOMIC_COUNTER_BUFFERS": 16,
    "GL_MAX_COMPUTE_SHADER_STORAGE_BLOCKS": 16,
    "GL_MAX_COMPUTE_TEXTURE_IMAGE_UNITS": 32,
    "GL_MAX_COMPUTE_UNIFORM_BLOCKS": 15,
    "GL_MAX_COMPUTE_UNIFORM_COMPONENTS": 16384,

> The purging is also very clever. It's not just the matter of using the uniform or not. It's about the actuall effect.

Aaah, I'll have a look into that as well... Maybe my code is not as efficient as I want to believe ;) I guess it's still a good idea to group them in a buffer, for efficiency, good practice, and portability...

**Comment by einarf** [comment]
It could also be a driver bug. It should be simple enough to make a dummy compute shader with a few large uniform arrays making sure they are actually used in the shader body.

Still it's good practice to stay below 1024 components for uniforms since this is the lower limit for the 4.3 spec. Using UBOS for the larger data arrays are not that difficult.

**Comment by LeoBosse** [comment]
> Still it's good practice to stay below 1024 components for uniforms since this is the lower limit for the 4.3 spec. Using UBOS for the larger data arrays are not that difficult.
For sure, I will do that. As I am a complete beginner with shaders, I just did the first thing I understood.

Now, it works, the main problem being between the chair and the computer. All the single float values that were disappearing were indeed not used in the shader code. So just deleting them solved it. BUT, the only array that disappeared was used in the code, and commenting it raised an error of undeclared variable. I noticed it was the only array with a small size (3), so I passed it as a vec3 instead, and it solved the problem!

So I believe this closes the issue :) Thanks a lot for your quick and detailed help!
